File: 78-subsets/subsets.cpp
class Solution {
public:
    void solve(vector<int> &n, vector<vector<int>> &ans, vector<int> &subset, int i){
        if (i == n.size()){
            ans.push_back(subset);
            return ;
        }

        // We can either include
        subset.push_back(n[i]);
        solve(n, ans, subset, i+1);
        // or not
        subset.pop_back();
        solve(n, ans, subset, i+1);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> all_subset;
        vector<int> subset;
        solve(nums, all_subset, subset, 0);

        return all_subset;
    }
};

/*
    - We will recursively create all the subset
*/

File: 78-subsets/subsets.go
func solve(nums []int, idx int, temp []int, ans *[][]int) {
    if idx == len(nums) {
        tmp := make([]int, len(temp))
        copy(tmp, temp)
        *ans = append(*ans, tmp)

        return
    }

    temp = append(temp, nums[idx])
    solve(nums, idx+1, temp, ans)
    temp = temp[:len(temp)-1]
    solve(nums, idx+1, temp, ans)
}

func subsets(nums []int) [][]int {
    var temp []int
    var ans [][]int

    solve(nums, 0, temp, &ans)

    return ans
}